Add Get-Uptime (C# code)

iSazonov opened the pull request:
1. Add Get-Uptime cmdlet (C# code)
2. Modify psd1 files
3. Add tests

Close #2497 

The PR replace #2507

iSazonov committed:
Add Get-Uptime (C# code)

1. Add Get-Uptime cmdlet (C# code)
2. Modify psd1 files
3. Add tests

msftclas commented:
Hi __@iSazonov__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, MSBOT;

lzybkr commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetUptime.cs:
```diff
@@ -0,0 +1,82 @@
+        /// <value></value>
+        [Parameter(ParameterSetName = "Timespan")]
+        [Parameter(ParameterSetName = "Since")]
+        public SwitchParameter Pretty { get; set; } = new SwitchParameter();
```
Do we need this parameter?

With strongly typed output, we can always use pretty formatting and still have useful output.

iSazonov replied:
My opinion: Get-Date has -Format parameter,  why not make the Pretty for Get-Uptime.

 @vors, lzybkr and daxian-dbw maybe you will have a short meeting and return with the conclusion?

lzybkr commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetUptime.cs:
```diff
@@ -0,0 +1,82 @@
+    /// This class implements Get-Uptime.
+    /// </summary>
+    [Cmdlet(VerbsCommon.Get, "Uptime", DefaultParameterSetName = "Timespan", HelpUri = "")]
+    public class GetUptimeCommand : PSCmdlet
```
The cmdlet should use the `OutputType` attribute.

iSazonov replied:
I agree. See next my comment.

iSazonov replied:
Fixed.

lzybkr commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetUptime.cs:
```diff
@@ -0,0 +1,82 @@
+        {
+            // Get-Uptime return null if IsHighResolution = false  
+            // because stopwatch.GetTimestamp() return DateTime.UtcNow.Ticks
+            // instead of ticks from system startup
```
It seems bad if this cmdlet silently doesn't work. At minimum, I'd expect an error, but really, why can't we find a way that always works?

iSazonov replied:
"silently" is from @daxian-dbw in #2507. I agree that user must get a error message about a problem. But in this case it is a question because such event is likely _very rare_ because all modern platforms should work correctly. On the other hand, the user can always change behavior using `-ErrorAction`  @daxian-dbw please comment too.

Windows has only one correct way to get uptime that is 'System Uptime' performance counter. But performance counters still are not ported. I could not find anything other.

daxian-dbw replied:
I take back my previous comment. Silently returning `null` is bad in the following example:
```
$timespan = [timespan]::new(24, 0, 0) # time span of 24 hours 
if ((Get-Uptime) -gt $timespan) {
    ## take the server offline for maintenance
}
```
In this case, the server may never get maintained 😦

iSazonov replied:
@daxian-dbw Thanks for great sample.
Clear. I'd add terminate throw (ThrowTerminatingError) after conclusion about Pretty.

daxian-dbw replied:
I think we can do something like this in the test:
```
It "test Get-Uptime" {
    if (System.Diagnostics.Stopwatch.IsHighResolution) {
        ## test the expected function of Get-Uptime
    } else {
        ## test the expected fully qualified name of the terminating error
    }
}
```

iSazonov replied:
@daxian-dbw Thanks for the idea. It seems that `else` option will never execute. Shouldn't we have a guaranteed check all options every time we run the test?
Is there any possibility of temporary overload code (System.Diagnostics.Stopwatch.IsHighResolution) for the test? (Load an assembly in Powershell?)

daxian-dbw replied:
I don't know if it's possible to overload `Stopwatch.IsHighResolution`, especially in the case of binary modules where powershell doesn't play a role of resolving types.
Test hooks may help you achieve this, take a look at [how it's used in Get-Help -Online](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/help/HelpCommands.cs#L623), and [its test](https://github.com/PowerShell/PowerShell/blob/master/test/powershell/engine/HelpSystem.OnlineHelp.Tests.ps1#L28).

However, I think using a test hook here may be an overkill -- we only throw a terminating error in this case and *maybe* a one-time manual check would be enough. @JamesWTruher, can you please weigh in?

iSazonov replied:
Since there was no feedback I added hook and test.

@lzybkr and @daxian-dbw  please continue the review.

lzybkr commented on src/Modules/Windows-Full/Microsoft.PowerShell.Utility/Microsoft.PowerShell.Utility.psd1:
```diff
@@ -2,7 +2,7 @@
 Author="Microsoft Corporation"
 CompanyName="Microsoft Corporation"
-Copyright="� Microsoft Corporation. All rights reserved."
+Copyright="� Microsoft Corporation. All rights reserved."
```
This change shouldn't be here.

iSazonov committed:
Revert typo

SteveL-MSFT commented:
@PowerShell/powershell-committee reviewed the -pretty parameter and agreed that we should use formatting/output to handle that and improve formatting of timespan rather than adding capability into specific cmdlets.

iSazonov committed:
Remove Pretty parameter after Powershell committee conclusion

1. Remove Pretty parameter
2. Remove tests for Pretty
3. Add Throw if  IsHighResolution is False
4. Fix formatting

iSazonov commented:
@SteveL-MSFT  Thanks for conclusion! I did the fix.

iSazonov committed:
Add test to check throw

1. Add hook that simulate 'System.Diagnostics.Stopwatch.IsHighResolution
is false' to test Get-Uptime throw
2. Add test to check throw

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetUptime.cs:
```diff
@@ -0,0 +1,71 @@
+            else
+            {
+                WriteDebug("System.Diagnostics.Stopwatch.IsHighResolution return 'False'.");
+                string msg = "Platform is not supported";
```
Error strings should be put in a `.resx` file so that it can be localized. Each cmdlet has its resource strings in a single `.resx` file, for example [WriteErrorStrings.resx](https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.Commands.Utility/resources/WriteErrorStrings.resx) only contains 1 resource string at [here](https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.Commands.Utility/resources/WriteErrorStrings.resx#L120), and it's used in [write.cs](https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.Commands.Utility/commands/utility/write.cs#L450). `Start-PSBuild -ResGen` will generate the C# binding for the `.resx` file.

iSazonov replied:
Fixed.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetUptime.cs:
```diff
@@ -0,0 +1,71 @@
+            {
+                WriteDebug("System.Diagnostics.Stopwatch.IsHighResolution return 'False'.");
+                string msg = "Platform is not supported";
+                Exception exc = new Exception(msg);
```
It should be a `NotSupportedException`, and the message should be more descriptive, for example, explain that `Stopwatch.IsHighResolution` is false on this system and thus system up time cannot be retrieved.

iSazonov replied:
Fixed.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetUptime.cs:
```diff
@@ -0,0 +1,71 @@
+                WriteDebug("System.Diagnostics.Stopwatch.IsHighResolution return 'False'.");
+                string msg = "Platform is not supported";
+                Exception exc = new Exception(msg);
+                ThrowTerminatingError(new ErrorRecord(exc, "UptimePlatformUnsupported", ErrorCategory.NotImplemented, null));
```
Note that, we usually use the id of the resource string in `.resx` file as the `errorId`.

daxian-dbw commented on src/System.Management.Automation/engine/Utils.cs:
```diff
@@ -1577,6 +1577,8 @@ public static class InternalTestHooks
         // It's useful to test that we don't depend on the ScriptBlock and AST objects and can use a re-parsed version.
         internal static bool IgnoreScriptBlockCache;
+        // Simulate 'System.Diagnostics.Stopwatch.IsHighResolution is false' to test Get-Uptime throw
+        internal static bool StopwatchIsHighResolutionIsFalse;
```
How about naming it `StopwatchIsNotHighResolution`? I feel the current variable name is a bit confusing.

iSazonov replied:
Fixed.

(Maybe `StopwatchIs_Not_HighResolution`?)

daxian-dbw replied:
I think `StopwatchIsNotHighResolution` is clear and descriptive enough 😄

iSazonov replied:
Clear. Closed.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetUptime.cs:
```diff
@@ -0,0 +1,71 @@
+        /// </summary>
+        /// <value></value>
+        [Parameter(ParameterSetName = "Timespan")]
+        public SwitchParameter Timespan { get; set; } = new SwitchParameter();
```
Fixed.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetUptime.cs:
```diff
@@ -0,0 +1,71 @@
+    /// <summary>
+    /// This class implements Get-Uptime
+    /// </summary>
+    [Cmdlet(VerbsCommon.Get, "Uptime", DefaultParameterSetName = "Timespan", HelpUri = "")]
```
Very good! 

Fixed.

In GetDateCommand.cs I found:
```powershell
    [OutputType(typeof(string), ParameterSetName = new string[] { "UFormat", "net" })]
    [OutputType(typeof(DateTime), ParameterSetName = new string[] { "net" })]
```
Should we do similarly?

daxian-dbw replied:
Good question! I searched the code base and it seems the `ParameterSetName` information is not really used anywhere today. But having more metadata is always a good thing and it might be useful someday, so yes let's include the parameter set names :)

As for the namespaces, my recommendation is to not include unnecessary namespace prefix unless it will result in ambiguity otherwise.

We will need to request a help uri for this new cmdlet. @Francisco-Gamino, is it OK to leave it as an empty string at this moment? Will this cause test failures in our nightly build?

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetUptime.cs:
```diff
@@ -0,0 +1,71 @@
+            // because stopwatch.GetTimestamp() return DateTime.UtcNow.Ticks
+            // instead of ticks from system startup.
+            // InternalTestHooks.StopwatchIsHighResolutionIsFalse is used as test hook.
+            if (System.Diagnostics.Stopwatch.IsHighResolution && !InternalTestHooks.StopwatchIsHighResolutionIsFalse)
```
Fixed.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetUptime.cs:
```diff
@@ -0,0 +1,71 @@
+                        break;
+                    case "Since":
+                        // return Datetime when the system started up
+                        WriteObject(System.DateTime.Now.Subtract(uptime));
```
Fixed.

daxian-dbw replied:
#Closed

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetUptime.cs:
```diff
@@ -0,0 +1,71 @@
+            }
+            else
+            {
+                WriteDebug("System.Diagnostics.Stopwatch.IsHighResolution return 'False'.");
```
Fixed.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetUptime.cs:
```diff
@@ -0,0 +1,71 @@
+            // InternalTestHooks.StopwatchIsHighResolutionIsFalse is used as test hook.
+            if (System.Diagnostics.Stopwatch.IsHighResolution && !InternalTestHooks.StopwatchIsHighResolutionIsFalse)
+            {
+                TimeSpan uptime = TimeSpan.FromSeconds(System.Diagnostics.Stopwatch.GetTimestamp()/System.Diagnostics.Stopwatch.Frequency);
```
Fixed.

daxian-dbw replied:
Thanks for the clarification! #Closed

iSazonov committed:
Fixes after review

1. Remove Timespan parameter
2. Remove unneeded tests
3. Fix localization
4. Add string consts
5. Fix throw

daxian-dbw reviewed (approved):
LGTM.
The only missing part is the update to `OutputType` and @Francisco-Gamino's input about `HelpUri = ""`.

Francisco-Gamino commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetUptime.cs:
```diff
@@ -0,0 +1,72 @@
+    /// <summary>
+    /// This class implements Get-Uptime
+    /// </summary>
+    [Cmdlet(VerbsCommon.Get, "Uptime", DefaultParameterSetName = TimespanParameterSet, HelpUri = "")]
```
Hi, since this cmdlet does not have a HelpUri, please add code to skip it in the Online help tests located at 
https://github.com/PowerShell/PowerShell/blob/master/test/powershell/engine/Help/HelpSystem.OnlineHelp.Tests.ps1

daxian-dbw replied:
> $cmdletList = Import-Csv $filePath -ea Stop

@Francisco-Gamino, `$cmdletList` is reading from the test data files, which don't have `Get-Uptime` in them, so it seems `Get-Uptime` will not be hit by this test. Can you please confirm?

daxian-dbw commented:
@iSazonov I talked to @Francisco-Gamino offline about the forward link help URI for this `Get-Uptime`. We will request for a `fwlink` internally, and once we get it, will update the code with it. For now, it's OK to leave the `HelpUri` a blank string.

So now the only missing part is the update to `OutputType`. Thanks!

iSazonov committed:
Fix OutputType

iSazonov commented:
OutputType is fixed.
@daxian-dbw thanks for great comments!

daxian-dbw commented:
LGTM. Thanks @iSazonov!

iSazonov commented:
@daxian-dbw Thanks for the review and usefull comments!
